#1.Download Dataset 

 -Install required libraries
 
 -Download data from kaggel
 
 -View Dataset files
 
 -Load training set with Pandas
 
 -Load testing set with Pandas
 

In [1]:
!pip install opendatasets

In [3]:
import opendatasets as od

##Download data from kaggle:

 We'll use the opendatasets library to download the dataset

 We'll need to upload our Kaggle API key(kaggle.json file) to do this

In [4]:
od.download('https://www.kaggle.com/competitions/new-york-city-taxi-fare-prediction/rules')

100%|██████████| 1.56G/1.56G [00:05<00:00, 335MB/s]



Extracting archive .\new-york-city-taxi-fare-prediction/new-york-city-taxi-fare-prediction.zip to .\new-york-city-taxi-fare-prediction


##View Dataset Files:

  Lets look at the size,no. of lines and first few lines of each files

In [3]:
import pandas as pd

In [4]:
train_data = pd.read_csv('train.csv')

In [3]:
len(train_data)

152977

In [5]:
test_data = pd.read_csv("test.csv")

In [5]:
len(test_data)

9914

The test data is significantly smaller than the train data

In [6]:
train_data.head()

key  fare_amount          pickup_datetime  \
0    2009-06-15 17:26:21.0000001          4.5  2009-06-15 17:26:21 UTC   
1    2010-01-05 16:52:16.0000002         16.9  2010-01-05 16:52:16 UTC   
2   2011-08-18 00:35:00.00000049          5.7  2011-08-18 00:35:00 UTC   
3    2012-04-21 04:30:42.0000001          7.7  2012-04-21 04:30:42 UTC   
4  2010-03-09 07:51:00.000000135          5.3  2010-03-09 07:51:00 UTC   

   pickup_longitude  pickup_latitude  dropoff_longitude  dropoff_latitude  \
0        -73.844311        40.721319         -73.841610         40.712278   
1        -74.016048        40.711303         -73.979268         40.782004   
2        -73.982738        40.761270         -73.991242         40.750562   
3        -73.987130        40.733143         -73.991567         40.758092   
4        -73.968095        40.768008         -73.956655         40.783762   

   passenger_count  
0              1.0  
1              1.0  
2              2.0  
3              1.0  
4              1.0

In [7]:
test_data.head()

key          pickup_datetime  pickup_longitude  \
0  2015-01-27 13:08:24.0000002  2015-01-27 13:08:24 UTC        -73.973320   
1  2015-01-27 13:08:24.0000003  2015-01-27 13:08:24 UTC        -73.986862   
2  2011-10-08 11:53:44.0000002  2011-10-08 11:53:44 UTC        -73.982524   
3  2012-12-01 21:12:12.0000002  2012-12-01 21:12:12 UTC        -73.981160   
4  2012-12-01 21:12:12.0000003  2012-12-01 21:12:12 UTC        -73.966046   

   pickup_latitude  dropoff_longitude  dropoff_latitude  passenger_count  
0        40.763805         -73.981430         40.743835                1  
1        40.719383         -73.998886         40.739201                1  
2        40.751260         -73.979654         40.746139                1  
3        40.767807         -73.990448         40.751635                1  
4        40.789775         -73.988565         40.744427                1

The test data seems to be missing the fare amount column,we can then infer that 'fare_amount' is the target attribute

Once the data has been evaluated it will be submitted into a submission file,an example will be displayed below

In [8]:
sub_data = pd.read_csv('sample_submission.csv')

In [9]:
sub_data.head()

key  fare_amount
0  2015-01-27 13:08:24.0000002        11.35
1  2015-01-27 13:08:24.0000003        11.35
2  2011-10-08 11:53:44.0000002        11.35
3  2012-12-01 21:12:12.0000002        11.35
4  2012-12-01 21:12:12.0000003        11.35

The evaluation metric for this submission is root-mean squared error(RMSE)

Our goal is a low score-the score is difference between the ground truth and the value of the prediction model

Lower scores are preferred

#Explore the Datasets:
 -Basic information about the training and testing sets
 -Exploratory Analysis and Visualization
 -Ask and Answer questions

In [10]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 152977 entries, 0 to 152976
Data columns (total 8 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   key                152977 non-null  object 
 1   fare_amount        152976 non-null  float64
 2   pickup_datetime    152976 non-null  object 
 3   pickup_longitude   152976 non-null  float64
 4   pickup_latitude    152976 non-null  float64
 5   dropoff_longitude  152975 non-null  float64
 6   dropoff_latitude   152975 non-null  float64
 7   passenger_count    152976 non-null  float64
dtypes: float64(6), object(2)
memory usage: 9.3+ MB


In [12]:
train_data.describe()

fare_amount  pickup_longitude  pickup_latitude  dropoff_longitude  \
count  152976.000000     152976.000000    152976.000000      152975.000000   
mean       11.360153        -72.502760        39.902041         -72.507053   
std         9.866959         10.577658        10.133938          10.407592   
min       -44.900000       -736.550000     -3116.285383         -84.654241   
25%         6.000000        -73.992036        40.734964         -73.991265   
50%         8.500000        -73.981732        40.752730         -73.980056   
75%        12.500000        -73.967000        40.767160         -73.963491   
max       500.000000         40.787575       401.083332          40.851027   

       dropoff_latitude  passenger_count  
count     152975.000000    152976.000000  
mean          39.928660         1.680682  
std            6.112461         1.305625  
min          -74.035197         0.000000  
25%           40.734108         1.000000  
50%           40.753243         1.000000  
75%           40.768081         2.000000  
max          404.616667         6.000000

In [16]:
train_data['pickup_datetime'] = pd.to_datetime(train_data['pickup_datetime'])

train_data['pickup_datetime'].min(), train_data['pickup_datetime'].max()

(Timestamp('2009-01-01 00:35:03+0000', tz='UTC'),
 Timestamp('2015-06-30 22:54:07+0000', tz='UTC'))

Observations about the train data:

 -15297 rows of data

 -There appears to be some missing values

 -The fare_amount ranges from $-44.9 to $5000

 -1 to 6 passengers

 -errors may be present in longitude and latitude values

 -Data has been collected for over 6 years(Jan 1st 2009 to June 30th 2015)

In [11]:
test_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9914 entries, 0 to 9913
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   key                9914 non-null   object 
 1   pickup_datetime    9914 non-null   object 
 2   pickup_longitude   9914 non-null   float64
 3   pickup_latitude    9914 non-null   float64
 4   dropoff_longitude  9914 non-null   float64
 5   dropoff_latitude   9914 non-null   float64
 6   passenger_count    9914 non-null   int64  
dtypes: float64(4), int64(1), object(2)
memory usage: 542.3+ KB


In [13]:
test_data.describe()

pickup_longitude  pickup_latitude  dropoff_longitude  dropoff_latitude  \
count       9914.000000      9914.000000        9914.000000       9914.000000   
mean         -73.974722        40.751041         -73.973657         40.751743   
std            0.042774         0.033541           0.039072          0.035435   
min          -74.252193        40.573143         -74.263242         40.568973   
25%          -73.992501        40.736125         -73.991247         40.735254   
50%          -73.982326        40.753051         -73.980015         40.754065   
75%          -73.968013        40.767113         -73.964059         40.768757   
max          -72.986532        41.709555         -72.990963         41.696683   

       passenger_count  
count      9914.000000  
mean          1.671273  
std           1.278747  
min           1.000000  
25%           1.000000  
50%           1.000000  
75%           2.000000  
max           6.000000

Our data appears to be exclusively numeric, to confirm this

In [64]:
test_data.isna()

key  pickup_datetime  pickup_longitude  pickup_latitude  \
0     False            False             False            False   
1     False            False             False            False   
2     False            False             False            False   
3     False            False             False            False   
4     False            False             False            False   
...     ...              ...               ...              ...   
9909  False            False             False            False   
9910  False            False             False            False   
9911  False            False             False            False   
9912  False            False             False            False   
9913  False            False             False            False   

      dropoff_longitude  dropoff_latitude  passenger_count  
0                 False             False            False  
1                 False             False            False  
2                 False             False            False  
3                 False             False            False  
4                 False             False            False  
...                 ...               ...              ...  
9909              False             False            False  
9910              False             False            False  
9911              False             False            False  
9912              False             False            False  
9913              False             False            False  

[9914 rows x 7 columns]

In [60]:
test_data['pickup_datetime'] = pd.to_datetime(test_data['pickup_datetime'])

test_data['pickup_datetime'].min(), test_data['pickup_datetime'].max()

(Timestamp('2009-01-01 11:04:24+0000', tz='UTC'),
 Timestamp('2015-06-30 20:03:50+0000', tz='UTC'))

Observations of the test data:

 -There are 9914 rows of data

 -There are no missing values

 -No obvious data entries 

 -Pick up dates range from Jan 1st 2009 to June 30th 2015(6 years)

 -We can use the ranges of the test data to drop outliers and invalid data from the training set

#Exploratory Data Analysis and Visualization

##Ask and answer Questions

1 What is the busiest day of the week

2 What is the busiest time of day

3 In which month are the fares busiest

4 Which pickup locations have the highest fares

5 Which drop off locations haves the  highest fares

6 What is the average ride distance

7

Performing EDA and asking questions helps us understand the data and prompts ideas for feature engineering

Its adviced to apply an iterative approach (do some EDA, feature engineering, train a  model and repeat)

#Prepare data set for training
 -Split training and validation set

 -fill/remove missing  values

 -Extract inputs and outputs
 
     -training

     -validation

     -test

##Split training and validation set:

We will set aside 20% of the data as the validation set(models are evaluated on unseen data)

Because the test and train data have the same range we can pick a random  20% fraction

In [7]:
from sklearn.model_selection import train_test_split

In [20]:
train_x, val_x = train_test_split(train_data, test_size=0.2, random_state=42)

In [21]:
len(train_x), len(val_x)

(2088979, 522245)

##Fill/Missing Values:
Our data sets have no missing values but if they did we would remove them rather than replace them as the data set is too large

In [22]:
train_x = train_x.dropna()
val_x = val_x.dropna()

#Extract Inputs and Outputs

In [23]:
train_x.columns

Index(['key', 'fare_amount', 'pickup_datetime', 'pickup_longitude',
       'pickup_latitude', 'dropoff_longitude', 'dropoff_latitude',
       'passenger_count'],
      dtype='object')

In [24]:
input_cols = ['pickup_datetime', 'pickup_longitude',
       'pickup_latitude', 'dropoff_longitude', 'dropoff_latitude',
       'passenger_count']

In [25]:
target_col = ['fare_amount']

##Training

In [26]:
train_inputs = train_x[input_cols]

In [27]:
train_target = train_x[target_col]

In [28]:
train_inputs

pickup_datetime  pickup_longitude  pickup_latitude  \
93357    2010-05-18 08:17:00 UTC        -73.990708        40.733003   
454068   2011-03-05 00:39:08 UTC        -73.989558        40.723098   
890949   2011-12-10 12:41:46 UTC        -73.989168        40.756114   
1432608  2009-07-17 17:23:00 UTC        -73.982283        40.778957   
2391898  2011-08-26 08:11:00 UTC        -73.789645        40.646953   
...                          ...               ...              ...   
110268   2012-03-17 01:34:48 UTC        -73.988959        40.719074   
1692743  2011-06-12 23:00:13 UTC        -73.958278        40.801008   
2356330  2013-07-15 15:12:00 UTC        -73.988052        40.753927   
2229084  2009-03-19 23:11:29 UTC        -73.990181        40.728880   
2219110  2011-04-02 11:06:08 UTC        -73.985506        40.759093   

         dropoff_longitude  dropoff_latitude  passenger_count  
93357           -73.979550         40.754852              5.0  
454068          -73.970929         40.793238              3.0  
890949          -73.975356         40.765202              2.0  
1432608         -73.967960         40.762510              1.0  
2391898         -73.969082         40.772753              3.0  
...                    ...               ...              ...  
110268          -73.978996         40.683716              1.0  
1692743         -73.969626         40.800258              1.0  
2356330         -73.987852         40.767122              2.0  
2229084         -73.998575         40.733901              1.0  
2219110         -73.975166         40.757697              2.0  

[2088960 rows x 6 columns]

In [29]:
train_target

fare_amount
93357            8.9
454068          17.3
890949           5.7
1432608          8.1
2391898         49.8
...              ...
110268          10.5
1692743          4.5
2356330          8.0
2229084          5.0
2219110          6.1

[2088960 rows x 1 columns]

##Validation

In [30]:
val_inputs = val_x[input_cols]
val_target = val_x[target_col]

In [31]:
val_inputs

pickup_datetime  pickup_longitude  pickup_latitude  \
482597   2012-03-30 12:24:00 UTC        -73.952863        40.785845   
743209   2011-06-14 23:22:00 UTC        -73.976523        40.755318   
2371738  2011-12-27 12:10:00 UTC        -74.004335        40.700000   
1455508  2011-02-16 22:55:10 UTC        -73.982677        40.750996   
2582289  2011-10-12 20:05:03 UTC        -73.989308        40.730454   
...                          ...               ...              ...   
2561049  2012-12-29 19:54:08 UTC        -73.982542        40.771566   
2276226  2010-02-14 07:23:32 UTC        -73.976658        40.788337   
2352991  2009-04-01 17:57:47 UTC        -73.981887        40.762965   
117966   2010-01-24 01:53:00 UTC        -74.006978        40.713488   
938939   2014-09-08 20:56:00 UTC        -73.982392        40.770625   

         dropoff_longitude  dropoff_latitude  passenger_count  
482597          -73.963192         40.781862              1.0  
743209          -73.999480         40.742548              2.0  
2371738         -73.971135         40.749187              1.0  
1455508         -73.987300         40.720453              1.0  
2582289         -74.002808         40.718356              1.0  
...                    ...               ...              ...  
2561049         -73.966520         40.753350              2.0  
2276226         -73.973097         40.764937              1.0  
2352991         -73.993880         40.761354              3.0  
117966          -73.980597         40.729630              1.0  
938939          -73.976652         40.792642              6.0  

[522244 rows x 6 columns]

In [32]:
val_target

fare_amount
482597           6.9
743209           7.3
2371738         11.7
1455508          8.9
2582289          7.7
...              ...
2561049         10.5
2276226          7.3
2352991          7.9
117966          11.3
938939          10.5

[522244 rows x 1 columns]

#Train Hardcoded and Baseline Models:

Create a simple hardcoded or basline model to establish the minimum score  that any proper ML model should beat

Hardcoded model: always predict average fare

Baseline model: Linear Regression

###Lets create a simple model that predicts the average fare

In [33]:
import numpy as np

In [34]:
class MeanRegressor:
    def fit(self, inputs, targets):
        self.mean = targets.mean()

    def predict(self, inputs):
        return np.full(inputs.shape[0], self.mean)

In [36]:
mean_model = MeanRegressor()

In [37]:
mean_model.fit(train_inputs, train_target)

In [38]:
mean_model.mean

fare_amount    11.350513
dtype: float64

In [40]:
train_prediction = mean_model.predict(train_inputs)

In [41]:
train_prediction

array([11.35051306, 11.35051306, 11.35051306, ..., 11.35051306,
       11.35051306, 11.35051306])

In [42]:
val_prediction = mean_model.predict(val_inputs)

In [43]:
val_prediction

array([11.35051306, 11.35051306, 11.35051306, ..., 11.35051306,
       11.35051306, 11.35051306])

Let us see how off we are by comparing our target to our prediction

In [44]:
train_target

fare_amount
93357            8.9
454068          17.3
890949           5.7
1432608          8.1
2391898         49.8
...              ...
110268          10.5
1692743          4.5
2356330          8.0
2229084          5.0
2219110          6.1

[2088960 rows x 1 columns]

In [45]:
train_prediction

array([11.35051306, 11.35051306, 11.35051306, ..., 11.35051306,
       11.35051306, 11.35051306])

We shall compare the using RMSE

In [46]:
from sklearn.metrics import mean_squared_error

In [48]:
def rmse(target, prediction):
    return mean_squared_error(target, prediction, squared=False)

In [49]:
train_rmse = rmse(train_target, train_prediction)
train_rmse

9.813927582987137

From this we can sumize that our predictions are off our taget by a value of 9(this is not a good value)

In [50]:
val_rmse = rmse(val_target, val_prediction)
val_rmse

9.913454595810046

This remains the same for the validation set

next we will:

##Train and Evaluate Baseline Model:we will train a linear regression model as our baseline, which tries to express the target as a weighted sum of the inputs

In [61]:
from sklearn.linear_model import LinearRegression

In [62]:
linear_model = LinearRegression()

In [66]:
print(train_inputs.dtypes)

pickup_datetime       object
pickup_longitude     float64
pickup_latitude      float64
dropoff_longitude    float64
dropoff_latitude     float64
passenger_count      float64
dtype: object


In [68]:
train_inputs['pickup_datetime'] = pd.to_datetime(train_inputs['pickup_datetime'])
train_inputs['year'] = train_inputs['pickup_datetime'].dt.year
train_inputs['month'] = train_inputs['pickup_datetime'].dt.month
train_inputs['day'] = train_inputs['pickup_datetime'].dt.day
train_inputs['hour'] = train_inputs['pickup_datetime'].dt.hour

# drop original timestamp column
train_inputs = train_inputs.drop(columns=['pickup_datetime'])

In [69]:
linear_model.fit(train_inputs, train_target)

LinearRegression()

##Observations

-This is a supervised linear regression problem

-The training data is 5.5GB

-The test set is much smaller(<10000 rows)

-The training set has the following 8 columns: 

        -key(unique identifer)
        -fare_amount(target column)
        -pickup_datetime
        -pickup_longitude
        -pickup_latitude
        -dropoff_longitude
        -dropoff_latitude
        -passenger_count

-The test set has all the columns except the target attribute(fare_amount)        

-The submission file should contain the key and fare_amount for each test sample